### PR TITLE
Move collate to data

### PR DESCRIPTION
**Pull request by RdoubleA**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

This should've happened a long time ago. Collaters are fundamentally a data helper and belong in `torchtune/data`. This needs to happen regardless of the approach taken in #1421. This is also discussed in #1261. 

#### Changelog
Move collate.py and update all callsites.

#### Test plan
Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.)

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Example of docstring: https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285
Example in our docs: https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models

- [ ] I did not change any public API;
- [ ] I have added an example to docs or docstrings;

### CODE REVIEW

File: tests/torchtune/data/test_collate.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 import torch
 
-from torchtune.utils.collate import padded_collate
+from torchtune.data._collate import padded_collate
```
Comment (SalmanMohammadi) [inline tests/torchtune/data/test_collate.py:11]: super nit: not sure if you need the full namespace
```suggestion
from torchtune.data import padded_collate
```

File: docs/source/tutorials/datasets.rst [hunk 2]
```diff
@@ -97,7 +97,7 @@ on Hugging Face's `documentation. <https://huggingface.co/docs/datasets/en/loadi
 Setting max sequence length
 ---------------------------
 
-The default collator, :func:`~torchtune.utils.padded_collate`, used in all
+The default collator, :func:`~torchtune.padded_collate`, used in all
```
Comment (SalmanMohammadi) [inline docs/source/tutorials/datasets.rst:100]:
```suggestion
The default collator, :func:`~torchtune.data.padded_collate`, used in all
```

**Comment by SalmanMohammadi** [comment]
Looks like there's a circular import in `collate.py`?

File: tests/torchtune/data/test_collate.py (hunk 1 above)
Comment (RdoubleA) [reply tests/torchtune/data/test_collate.py:11]: I am totally burned by all the circular imports, so I'm going to be explicit in most cases
